### PR TITLE
Add an initial architecture document

### LINKED ISSUE #50: Complete Architecture
Need to have a complete architecture document. As of now we have some pictures and slides, 
but the knowledge of the architecture (components, communication mechanism) are distributed and not documented in a central place.

### PR DESCRIPTION
**Pull request by berezovskyi**
Closes #50 

- [ ] robots pick product from shelves, and deliver them to conveyor belts, and not the other way around as it is represented in the diagrams

### CODE REVIEW

**Comment by berezovskyi** [comment]
Please feel free to just push corrections and additions (I think a lot of material is missing in regard to the planning arch, ontologies, safety, twin/robot architecture) to the branch directly.

**Comment by KRaizer** [comment]
Thanks for the good work. We will take a closer look at it as soon as possible.
From a quick look, one thing that requires change is that robots pick product from shelves, and deliver them to conveyor belts, and not the other way around as it is represented in the diagrams.
The reason for this is that in our scenario the products are already available at the warehouse, which works as a distribution node, and they have to be shipped and delivered to the clients by trucks.

**Comment by kicacacko** [comment]
I reviewed the architecture and made small change on the Integration layer to Adaptation Layer
I think 5. is missing in the figure.
Actually I see two steps missing:
a) the information going out from the capability metadata services and giving capability information to the warehouse controller
b) the warehouse controller publishes configuration information for the planning problem and planning domain service to access it. Btw what is configuration information? Is it which objects are active and the outline of the warehouse, OR it includes also capability information of the dynamic objects?

The Plan interpreter is something that in the earlier architecture we called the Estimator, right?

Could you provide an example for an Costum protocol/API between a service and an adapter. For example in the Planning Service, is it PDDL the protocol, the need for a transformation from RDF to PDDL, or I am missunderstanding something here.

**Comment by swarupmohalik** [comment]
On a superficial note, wouldn’t it be good to have a uniform knowledge base interface through, say a REST API, for all the services and devices, instead of adaptors which will proliferate the software components in the system?

The inputs for knowledge store requires the ontology defn and the instance.
The API can parse the instance once the URI for the definition is known.
The services can be replicated when necessary.

**Comment by berezovskyi** [comment]
@KRaizer thank you, I added the task for myself to edit the diagram and reflect needed changes.

@eanevul thanks  for the changes and the notes:

- step 5 is indeed missing, I mentally counted through it between the PDDL Generator and PDDL Engine
- I assumed that response to the capability metadata request is immediate but then I don't show the response from the Problem service to the Planning service.
- we need to define the configuration information, it is missing for now (see #53)
- Plan Interpreter is a new component that is supposed to track plan execution at runtime.
- yes, custom protocol/API/format is exactly PDDL in the case you mentioned. also the twin & ROS communication via e.g. ROS `common_msgs` is described by that arrow.

@swarupmohalik there will be a uniform interface based on REST and RDF for each service – this is exactly what OSLC is about. Centralising the knowledgebase, however, might be tricky because it will limit the scalability of the system. In order to simplify things, I will set up a TRS client collecting data from every service in the warehouse and updating a central triplestore accordingly. Therefore, a read-only SPARQL interface can be considered a "uniform knowledge base interface"?

As per these points:

_The inputs for knowledge store requires the ontology defn and the instance._ Exactly, Lyo "inputs" are called OSLC services (this is why we all all OSLC servces _adaptors_ to avoid confusion). "Ontology definitions" are done using RDFS vocabularies and OSLC shapes. Instances of the RDF data encoded in eg Turle that adhere to an OSLC shape are sent to the OSLC services.

_The API can parse the instance once the URI for the definition is known._ Generated Lyo code parses the serialised RDF from Turtle or other format and our REST handler methods receive Java POJO instances (under the hood we use Java annotations & reflection to do a bit of magic). 

_The services can be replicated when necessary._ Not sure I understand you correctly, but OSLC adaptors are just Java REST services, can be scaled horizontally in k8s if we implement triplestore communication safely. Also, a single domain spec (eg Warehouse DS) can be used across multiple adaptors and code for the domain can be generated into a single library that will be reused across REST services.

**Comment by berezovskyi** [comment]
ping @konstantino

File: doc/architecture.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+> I think a single master will conflict with the idea of a twin per device.
+
+> If we have a single master, the architecture figure needs to be updated to
+> reflect that.
+
+The devices communicate with the rest of the sandbox through their _twins_. The
+devices register in the sandbox through the _warehouse controller_ (there was an
+idea to have a standalone _registration & discovery service_ to deal with the
+proprietary discovery protocols by Leo, I think) and by reporting their
```
Comment (likelion) [inline doc/architecture.md:73]: Do you mean the OSGi thingy that uses UPnP to discover devices?

Comment (berezovskyi) [reply doc/architecture.md:73]: No, I has nothing specific in mind here. We were just discussing that robots might have a proprietary registration mechanism and we might want to have a separate service. Not sure if we are putting it into the 1st yr. though.

**Review by likelion — APPROVED** [review]
Ok as a working document.

**Pull request merged**